SingularityService/src/main/java/com/hubspot/singularity/data/StateManager.java

```diff
@@ -322,6 +328,22 @@ public SingularityState generateState(boolean includeRequestIds) {
     return numTasks.toCountMap();
   }
 
+  private ImmutablePair<List<SingularityPendingTaskId>, List<SingularityPendingTaskId>> filterOnDemandRequests (List<SingularityPendingTaskId> lateTasks) {
+    List<SingularityPendingTaskId> onDemandLateTasks = new ArrayList<>();
+    List<SingularityPendingTaskId> nonOnDemandLateTasks = new ArrayList<>();
+
+    List<String> requestIds = lateTasks.stream().map(lateTask -> lateTask.getRequestId()).collect(Collectors.toList());
+    List<RequestType> requestTypes = requestManager.getRequests(requestIds).stream().map(s -> s.getRequest().getRequestType()).collect(Collectors.toList());
+    for (int i = 0; i < lateTasks.size(); i++) {
+      if (requestTypes.get(i).equals(RequestType.ON_DEMAND)){
```

Comment:
Hm, I'm not sure this'll work as intended. For example, if we have two late tasks from the same request ID (so `requestIds` would be `[requestIdA, requestIdA]`), I think `requestManager.getRequests([requestIdA, requestIdA])` would just return `[requestA]`. Then the `requestTypes.get(i)` indexing wouldn't work as expected.

Reply:
An alternative could be something like this:

```java
Map<Boolean, List<SingularityPendingTaskId>> lateTasksPartitionedByOnDemand = lateTasks.stream()
  .collect(
    Collectors.partitioningBy(lateTask -> requestManager.getRequest(lateTask.getRequestId()).getRequestType() == RequestType.ON_DEMAND)
  );

// lateTasksPartitionedByOnDemand.get(true) == on demand late tasks
// lateTasksPartitionedByOnDemand.get(false) == everything else
```